### PR TITLE
Refactor some asset code

### PR DESCRIPTION
**Pull request by benfrankel**
Fixes https://github.com/TheBevyFlock/bevy_quickstart/issues/168.

### CODE REVIEW

**Review by paul-hansen — APPROVED** [review]
Looks great! I like the generic AssetMap, the name is much better too, the previous naming (e.g. ImageAsset) was confusing to use when we already have `Assets<Image>` as variable names become too similar.
Thanks!

File: src/game/assets.rs
Comment (janhohenheim) [inline src/game/assets.rs:110]:
```suggestion
        self.values()
```

File: src/game/assets.rs
Comment (janhohenheim) [inline src/game/assets.rs:105]: I'd implement `From<K, V, const N: usize> for AssetMap`, make `load` return `impl Into<AssetMap>` and then make the `load` implementations return the array directly. Less boilerplate that way.

File: src/game/assets.rs
Comment (janhohenheim) [inline src/game/assets.rs:97]: I get why this is called `Key`, but it highlights the implementation detail and not the intention behind it. Also, the function `load` doesn't map semantic sense on a key and implies this is conceptually the asset itself. Maybe something like `PreloadedAsset` would be nicer?

File: src/game/assets.rs
Comment (janhohenheim) [inline src/game/assets.rs:17]: Implement `FromWorld` for all `T: AssetKey + Resource` and these can be simple `init_resource` calls.

File: src/screen/loading.rs
Comment (janhohenheim) [inline src/screen/loading.rs:37]: Would be cool to use polymorphism here to reduce duplication, but I don't think that's possible here, is it? Some thing like

```rust
[image_map as IdkWhichType, sfx_map, soundtrack_map].iter().all(...)
```

File: src/game/spawn/player.rs
Comment (janhohenheim) [inline src/game/spawn/player.rs:29]: Given that we also use a sprite atlas in the context of a player, which is also a kind of "image map", this is some potentially very confusing wording. Consider finding a better name for `AssetMap` in general.

**Review by janhohenheim — CHANGES_REQUESTED** [review]

**Comment by janhohenheim** [comment]
Change the design doc accordingly after the review is through, but before merging.

File: src/game/assets.rs
Comment (benfrankel) [reply src/game/assets.rs:17]: It'd be `FromWorld` for `AssetMap<K>`, and this would become `app.init_resource::<AssetMap<ImageKey>>()`. Which is reasonable.

File: src/game/assets.rs
Comment (janhohenheim) [reply src/game/assets.rs:97]: @paul-hansen do you have an opinion here?

File: src/game/spawn/player.rs
Comment (janhohenheim) [reply src/game/spawn/player.rs:29]: @paul-hansen do you have an opinion here?

Comment (benfrankel) [reply src/game/spawn/player.rs:29]: How about `image_handles: Res<HandleMap<ImageKey>>`?

File: src/screen/loading.rs
Comment (benfrankel) [reply src/screen/loading.rs:37]: I did think about this. I'm not sure there's a more elegant option than just calling `.all_loaded` on each one, without ballooning the complexity. If there is a way, it should be a separate PR probably.

File: src/game/assets.rs
Comment (janhohenheim) [reply src/game/assets.rs:93]: I think that, fundamentally, actual keys are not a thing that loads anything

File: src/game/spawn/player.rs
Comment (janhohenheim) [reply src/game/spawn/player.rs:29]: Works for me

File: src/game/assets.rs
Comment (benfrankel) [reply src/game/assets.rs:93]: This is going to be a tradeoff between conceptual purity and code quality.

Comment (benfrankel) [reply src/game/assets.rs:105]: I pushed a potential workaround for this, removing `AssetKey::load` in favor of manual impls of `FromWorld` for `AssetMap<XyzKey>`. It adds some boilerplate for each new asset collection, but you get the conceptual purity.

Comment (janhohenheim) [reply src/game/assets.rs:105]: My argument stems from the idea that `Key` is a bit of an implementation detail anyways and does not help you conceptualize what is happening, hence `PreloadedAsset`.
I'm fine with leaving it like this if you disagree, especially considering @paul-hansen's endorsement as an unbiased third party.

Comment (benfrankel) [reply src/game/assets.rs:105]: I'd argue that the preloaded part is the implementation detail that the rest of the codebase doesn't need to know or care about. :P

File: src/game/assets.rs
Comment (janhohenheim) [inline src/game/assets.rs:105]: I think this trait is no longer needed

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/game/spawn/player.rs
Comment (paul-hansen) [reply src/game/spawn/player.rs:29]: No strong opinions here, since it's just a template so we can always rename to whatever we want.
I think making it clearer that it's handles not the actual assets might help some new people with the concepts slightly so I'm on board with that.

File: src/game/assets.rs
Comment (benfrankel) [reply src/game/assets.rs:105]: It's needed so that `AssetMap` (or `HandleMap`) only requires one generic.

File: src/game/spawn/player.rs
Comment (benfrankel) [reply src/game/spawn/player.rs]: Renamed to `HandleMap`.

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Looks good now! Will approve after the design doc is updated

**Comment by benfrankel** [comment]
@janhohenheim Updated design doc.

**Review by janhohenheim — APPROVED** [review]

**Pull request merged**
